build/gengo.bzl

```diff
@@ -0,0 +1,171 @@
+#           name="deepcopy-sources",
+#           base="zz_generated.deepcopy.go",
+#           tool="//vendor/k8s.io/code-generator/cmd/deepcopy-gen",
+#           match="+k8s:deepcopy-gen=",
+#           flags="--bounding-dirs k8s.io/kubernetes,k8s.io/api",
+#           packages=[
+#             "k8s.io/kubernetes/every/package",
+#             "k8s.io/kubernetes/that/wants",
+#             "k8s.io/kubernetes/a/zz_genrated.deepcopy.go/file",
+#           ],
+#           deps=[
+#             "vendor/github.com/every/package",
+#             "vendor/whatever.io/used/by/packages",
+#           ],
+#      )
+#   2) For every package in the packages list, add the following:
+#       load("//build:gengo.bzl", "k8s_gengo")
+#       k8s_gengo(name="generate-deepcopy", outs=["zz_generated.deepcopy.go"])
+#   3) Be sure to use the //build:go.bzl wrapped version of go_library
+#       load("@io_bazel_rules_go//go:def.bzl", "go_library")  # NO: generate-deepcopy fails
+#       load("//build:go.bzl", "go_library")  # YES: generate-deepcopy succeeds
+#   4) gazelle will now notice the existence of this file and add it to srcs
+#       go_library(
+#           name="go_default_library",
+#           srcs=[..., "zz_generated.deepcopy.go"],
+#           ...
+#       )
+
+load("@io_kubernetes_build//defs:go.bzl", _go_genrule="go_genrule")
+
+def _generate_prefix(tool, tool_flags, out, match, header):
+  """Set the variables that change for different generators.
+
+  Args:
+    tool: a gengo tool, something like k8s.io/code-generator/cmd/deepcopy-gen
+    tool_flags: unique flags for the tool, something like --bounding-dirs k8s.io/kubernetes
+    out: the basename of the file, something like zz_generated.deepcopy.go
+    match: search packages for files with this tag, something like +k8s:deepcopy-gen=
+    header: boilerplate file, something like vendor/k8s.io/code-generator/hack/boilerplate.go.txt
+  """
+  return """
+  tool='$(location {tool})'
+  tool_flags='{flags}'
+  out='{out}'
+  match='{match}'
+  head='{header}'
+""".format(tool=tool, tool_flags=tool_flags, out=out, match=match)
+
+def _generate_outs(packages):
+  """Return the move-out commands which copy files to where bazel expects."""
+  return "\n".join(["move-out '%s'" % p for p in packages])
+
+def source_files(pkgs):
+  """Return a list of sourcefiles targets for the specified packages.
+
+  The :go_default_library.sourcefiles target includes all the
+  non-generated *.go files for the package. Both the gengo and go list
+  tools use these files to reflect over the package.
+  """
+  return ["//%s:go_default_library.sourcefiles" % p for p in pkgs]
+
+def k8s_gengo_all(name, base, tool, flags, match, packages, deps):
+  """Use TOOL to generate BASE files in all the PACKAGES with a MATCH comment."""
+  # TODO(fejta): gazelle doesn't appear to auto-generate rules for this package
+  packages = [p for p in packages if not p.startswith("vendor/k8s.io/code-generator/_examples/")]
```

Comment:
this seems like a code smell. can't we filter this out before calling `k8s_gengo_all`?

Reply:
Done